interfaces: posix-mq: add new interface

IsaacJT opened the pull request:
This adds support for a new posix-mq interface, including AppArmor and seccomp rules. This allows creating, sending, and receiving IPC messages over POSIX message queues between snaps.

Support in AppArmor is not yet available, but it will be coming soon. The AppAmor rules in this interface are based on pre-release information from the developers implementing the feature.

AppArmor rule syntax:

```
['audit'] ('deny'|'allow'|)  'mqueue' [<PERMS>] [<OPTIONS>] [<NAME>] ','

PERMS = (SHORTPERMS+ | '(' <LONGPERMS_LIST> ')' )
SHORTPERMS = ('r'|'w')
LONGPERMS_LIST = <LONGPERM> ([,] <LONGPERM>)*
<LONGPERM> = ('read'|'write'|'create'|'open'|'delete'|'getattr'|'setattr')
OPTIONS = ('label='<LABEL_NAME>|'type='<TYPE>)
NAME = /.* 
```

alexmurray commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,237 @@
+        - app
+        - core
+    deny-auto-connection: false
+`
```
Can you also add a `posixMQBaseDeclarationPlugs` which looks something like the following? This should allow snaps from the same publisher to auto-connect.

```
const posixMQBaseDeclarationPlugs = `
  posix-mq:
    allow-installation: true
    allow-connection:
      slot-attributes:
        posix-mq: $PLUG(posix-mq)
    allow-auto-connection:
      slot-publisher-id:
        - $PLUG_PUBLISHER_ID
      slot-attributes:
        posix-mq: $PLUG(posix-mq)
`
```

alexmurray commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,237 @@
+}
+
+func (iface *posixMQInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	if path, err := iface.getPath(slot, slot.Name); err == nil {
```
Can you please add the permissions validation here as well?

mardy commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,237 @@
+      slot-snap-type:
+        - app
+        - core
+    deny-auto-connection: false
```
We probably want to be more restrictive on this, otherwise a snap could be able to impersonate another snap and read/write messages which were not meant for it.

Using a declaration similar to that of the shared-memory interface might make sense here too.

mardy commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,237 @@
+	"send", /* alias for write */
+	"create",
+	"delete",
+	"destroy", /* alias for delete */
```
why do we need the aliases?

IsaacJT replied:
I'm not sure - they were part of the prerelease AppArmor spec. I can remove them to clean it up a bit, if it's preferred to have fewer

mardy replied:
Let's wait and see what @pedronis says :-)

pedronis replied:
yes, unless we break something that shipped, we definitely don't want to support evolutionary leftovers

IsaacJT replied:
ACK - will remove the aliases

alexmurray commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,237 @@
+func (iface *posixMQInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:           posixMQSummary,
+		ImplicitOnCore:    true,
```
Does it make sense for `ImplicitOnCore` or `ImplicitOnClassic`? My understanding is we would always want a matching slot for a given plug and so there would be no implicit version of this interface.

IsaacJT replied:
Removed

alexmurray commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,237 @@
+
+func (iface *posixMQInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:           posixMQSummary,
```
Please add `BaseDeclarationSlots: posixMQBaseDeclarationSlots,`

alexmurray commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,237 @@
+	if perms, err := iface.getPermissions(slot, slot.Name()); err == nil {
+		// always allow these functions
+		var syscalls = []string{
+			"mq_notify",
```
Should `mq_notify` go with `mq_read`? ie. we should only allow `mq_notify` if we also allow `mq_read` since both allow to get information about what is in the queue.

mardy reviewed (commented):
Thanks Isaac!

I left a few comments inline.

alexmurray commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,237 @@
+			case "open":
+				fallthrough
+			case "create":
+				syscalls = append(syscalls, "mq_open")
```
`mq_open` is already in `syscalls` above so why add it a second time?

IsaacJT replied:
Good catch - removed

alexmurray reviewed (changes requested):
This is a good start - the other missing piece at this time would be to update `probeParserFeatures()` in `sandbox/apparmor/apparmor.go` to allow snapd to detect if the host apparmor supports `mqueue` - in which case we should then add a new feature called `mqueue` - ie something like:

```
	if tryAppArmorParserFeature(parser, "mqueue,") {
		features = append(features, "mqueue")
	}
```

Then in `BeforePrepareSlot()` and `BeforePreparePlug()` include a section at the top which looks something like (or perhaps this should be encapsulated into it's own function):

```
	if apparmor_sandbox.ProbedLevel() == apparmor_sandbox.Unsupported {
		// no apparmor means we don't have to deal with parser features
		return nil
	}
	features, err := apparmor_sandbox.ParserFeatures()
	if err != nil {
		return err
	}

	if !strutil.ListContains(features, "mqueue") {
		// then the host system doesn't have the required feature to compile the
		// policy
		return fmt.Errorf("cannot prepare slot / connect plug on system without mqueue support")
	}
```

IsaacJT commented:
Thanks @alexmurray  and @mardy, I've pushed further changes according to your comments. 

The unit tests will likely fail here as the interface now checks if the feature is supported.

alexmurray commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,291 @@
+		if path, ok := pathAttr.(string); ok {
+			return path, nil
+		}
+		return "", fmt.Errorf(`posix-mq slot %s "path" attribute must be a string, not %v`, name, pathAttr)
```
Can you please add validation using `cleanSubPath()` here as well?

alexmurray commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,291 @@
+		return fmt.Errorf("posix-mq path must conform to the POSIX message queue name specifications (see `man mq_overview`)")
+	} else {
+		return err
+	}
```
Can this path also please be validated using `apparmor_sandbox.ValidateNoAppArmorRegexp()`? In general I think it would be useful to structure the checks similarly to say how the currently-in-development `pkcs11` interface is over in https://github.com/snapcore/snapd/pull/11540

alexmurray commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,291 @@
+	}
+}
+
+func (iface *posixMQInterface) AutoConnect(plug *snap.PlugInfo, slot *snap.SlotInfo) bool {
```
Is this `AutoConnect()` required?

alexmurray commented on sandbox/apparmor/apparmor.go:
```diff
@@ -346,6 +346,9 @@ func probeParserFeatures() ([]string, error) {
 	if tryAppArmorParserFeature(parser, "capability audit_read,") {
 		features = append(features, "cap-audit-read")
 	}
+	if tryAppArmorParserFeature(parser, "mqueue,") {
```
The unit tests for `tryAppArmorParserFeature()` will also need to be updated.

alexmurray reviewed (changes requested):
Thanks for the updates - can you please ensure the mq path is validated via `apparmor_sandbox.ValidateNoAppArmorRegexp()`?

mardy commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,291 @@
+	if !implicitSystemPermanentSlot(slot) {
+		if path, err := iface.getPath(slot, slot.Name); err == nil {
+			spec.AddSnippet(fmt.Sprintf(`# POSIX Message Queue management
+mqueue (create delete getattr setattr read write) %s,
```
Sorry for not noticing it before, but all these paths should be quoted, otherwise bad things might happen if the path contains spaces and commas :-) :
```suggestion
mqueue (create delete getattr setattr read write) "%s",
```

The `ValidateNoAppArmorRegexp()` function will ensure that path does not contain a `"`, so we don't need to worry about escaping stuff.

IsaacJT commented:
Hi all, thanks for the new suggestions - I've added the changes to the path validation and unit tests

pedronis commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,237 @@
+
+const posixMQSummary = `allows access to POSIX message queues`
+
+const posixMQBaseDeclarationSlots = `
```
@alexmurray if I understand correctly this has a flat namespace like shared-memory so it needs to be superprivileged so that we can manage the shared namespace

IsaacJT replied:
ACK - will remove

IsaacJT replied:
@pedronis should both the plug and slot be set to `allow-installation: false` or just the slot, like in shared-memory?

pedronis replied:
I think the reasoning for shared-memory is the same here

alexmurray replied:
@pedronis thanks for the hint re managing the namespace - I hadn't thought of that - I wonder if we should document this somewhere (ie. interfaces which provide device ownership *or* which have a shared namespace should be super-privileged) or somesuch?

pedronis reviewed (commented):
couple of comments

IsaacJT commented:
Hi all,

I've rebased to the latest master and added an additional commit which uses the name of the slot as the path if no path has been given.

IsaacJT commented:
I've also removed get/setattr as they are not supported by the AppArmor rules, and I've removed the AppArmorConnectedSlot function as it was redundant (all permissions are provided by AppArmorPermanentSlot)

alexmurray commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,297 @@
+	if !implicitSystemPermanentSlot(slot) {
+		if path, err := iface.getPath(slot, slot.Name); err == nil {
+			spec.AddSnippet(fmt.Sprintf(`  # POSIX Message Queue management
+  mqueue (create delete read write) "%s",
```
Instead of hard-coding these permissions here would it be better to just use all of the ones listed in `posixMQPlugPermissions`?

IsaacJT replied:
Good point - I've changed that in the latest commit

alexmurray reviewed (approved):
LGTM - would be good to get @pedronis to review the plug/slot base declarations as I am still a bit unsure about that.

IsaacJT commented:
I added a new commit containing a suggestion from @alexmurray and corrected the unit tests to reflect it

mardy commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,300 @@
+}
+
+func (iface *posixMQInterface) isValidPermission(perm string) bool {
+	for _, validPerm := range posixMQPlugPermissions {
```
We have `strutil.ListContains()`

IsaacJT replied:
Fixed - thanks

mardy commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,300 @@
+func (iface *posixMQInterface) validatePermissionList(perms []string, name string) error {
+	for _, perm := range perms {
+		if !iface.isValidPermission(perm) {
+			return fmt.Errorf("posix-mq slot %s permission \"%s\" not valid, must be one of %v", name, perm, posixMQPlugPermissions)
```
IIRC, in another review @pedronis suggested not to print the slot name in the error messages.

IsaacJT replied:
Should I remove the names?

pedronis replied:
yes, higher levels will add the slot name to the error.  Also this and many other error paths are not tested.

mardy reviewed (approved):
Thanks Isaac, this looks good to me! I've justa dded nitpick comments and questions, but maybe before making further changes it's better to see what others have to say.

pedronis commented on tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml:
```diff
@@ -353,6 +353,9 @@ apps:
     plugs: [ polkit ]
+  posix-mq:
+    command: bin/run
+    plugs: [ posix-mq ]
```
is this enough? also won't this give us trouble as the feature is parser dependent?

IsaacJT replied:
@mvo5 added this - I'm not sure

mvo5 replied:
It was fine in my local run but maybe I did something else wrong. In any case it unblocks spread enough that we see some more failures from the static tests, notably https://github.com/snapcore/snapd/runs/5885786791?check_suite_focus=true#step:19:255 which looks like some mocking is missing and a bit below https://github.com/snapcore/snapd/runs/5885786791?check_suite_focus=true#step:19:283

IsaacJT replied:
I added a call to `apparmor_sandbox.MockFeatures()` so it no longer fails the tests due to the missing feature in `apparmor_parser`

mvo5 commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,300 @@
+	}
+
+	if !strutil.ListContains(features, "mqueue") {
+		return fmt.Errorf("AppArmor does not support POSIX message queues - cannot setup or connect interfaces")
```
My feeling is that the error is appropriate, the expectation of the user is to have mediation for mqueue and we cannot provide it. Yes, in the extreme case that there is no apparmor at all we move to devmode automatically (which is something we may need to revisit) - but here I think it would be surprising if everything is mediated but appamor is not because the userland is missing. Maybe a slightly more information message though? Something like:
```
cannot setup or connect interface: your AppArmor does not support POSIX message queue, see http://forum.snapcraft.io/t/29482
```
(I created a unlisted forum topic that explain things a bit, probably best if that gets expanded a little once we are closer to the feature. it could list for example a PPA).

pedronis commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,291 @@
+      slot-publisher-id:
+        - $PLUG_PUBLISHER_ID
+      slot-attributes:
+        posix-mq: $PLUG(posix-mq)
```
afaict there is no code atm checking that posix-mq is set to a string, or defaulting it to the slot/plug name as we do for content and shared-memory

IsaacJT replied:
If I understand you correctly this is already in the getPath function:

https://github.com/snapcore/snapd/pull/11590/files/4d14c3028cb30aece59556dcf696332a13d3e906#diff-81c542b03cd89ab9e295c71416ac4d48c89b02af12f513d203fe313e7d89aa2fR151

pedronis replied:
no,   the attribue referred here is called "posix-mq", it's labeling the slot to match up with consuming plugs (I imagine). It's not "path". I suggest looking how shared_memory.go deals with the labeling "shared-memory" attribute vs a "path" attribute.

pedronis commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,291 @@
+func (iface *posixMQInterface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
+	if !implicitSystemPermanentSlot(slot) {
+		if path, err := iface.getPath(slot, slot.Name); err == nil {
+			aaPerms := strings.Join(posixMQPlugPermissions, " ")
```
this probably merits a comment about the slot side getting all the permissions

pedronis reviewed (changes requested):
this needs more work. there is no validation of the posix-mq label attribute. Also missing unit tests for the (validation) error code paths

IsaacJT commented:
@pedronis I've made the changes you suggested and added more failure unit tests

pedronis commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,331 @@
+	if err := iface.checkPosixMQAppArmorSupport(); err != nil {
+		return err
+	}
+
```
this needs to validate posix-mq and default it to the plug name if it's missing (see similar code for content.go BeforePreparePlug)

pedronis reviewed (commented):
needs a bit more work

IsaacJT commented:
@pedronis I've added the `posix-mq` validation to `BeforePreparePlug` and also added some more unit tests to ensure the label is parsed/not parsed correctly

mvo5 reviewed (approved):
Samuele looked at it via my laptop and he is fine with it.

mvo5 reviewed (commented):
I added a bunch of comments, all nitpicks and suggestions, please don't feel bad by the amount of comments. The code is nice and well tested and I'm happy to land it but it would be nice to have a followup. But I will land this so that we can branch 2.56 and get it into beta.

alexmurray commented:
FWIW I would like to see @mvo5's suggested changes adopted before this is merged as they make it all a bit easier to read IMO.

IsaacJT commented on interfaces/builtin/posix_mq.go:
```diff
@@ -0,0 +1,344 @@
+			case "read":
+				syscalls = append(syscalls, "mq_timedreceive")
+				syscalls = append(syscalls, "mq_notify")
+				break
```
Very much a C habit :D thanks

IsaacJT commented:
@alexmurray @mvo5 I've made a number of improvements - thanks @mvo5 for the suggestions and tips!

Do you want me to squash the commits?

mvo5 commented:
Just fyi - here is my messing around https://paste.ubuntu.com/p/8ks72S9WyV/ with this, maybe it's useful (but all nitpicks).

> @alexmurray @mvo5 I've made a number of improvements - thanks @mvo5 for the suggestions and tips!
> 
> Do you want me to squash the commits?

Thank you,

mvo5 commented:
Thank you again! I merged it and also cherry picked for 2.55 so if we do another 2.55.5 release this will be part of it
